PolyEngine/Engine/Src/ECS/World.hpp

```diff
@@ -121,16 +121,61 @@ namespace Poly {
 				return **this;
 			}
 
-			ComponentIterator& operator++() { ++primary_iter; return *this; }
-			ComponentIterator operator++(int) { ComponentIterator ret(primary_iter); ++primary_iter; return ret; }
-			ComponentIterator& operator--() { --primary_iter; return *this; }
-			ComponentIterator operator--(int) { ComponentIterator ret(primary_iter); --primary_iter; return ret; }
+			ComponentIterator& operator++() { Increment(); return *this; }
+			ComponentIterator operator++(int) { ComponentIterator ret(primary_iter); Increment(); return ret; }
+			ComponentIterator& operator--() { Decrement(); return *this; }
+			ComponentIterator operator--(int) { ComponentIterator ret(primary_iter); Decrement(); return ret; }
+
+		private:
+			template<int zero = 0>
+			bool HasComponents(const Entity* entity) const
+			{
+				return true;
+			}
+			template<typename Component, typename... Rest>
+			bool HasComponents(const Entity* entity) const
+			{
+				if (entity->template GetComponent<Component>())
+					return HasComponents<Rest...>(entity);
+				else
+					return false;
+			}
+			void Increment()
+			{
+				do {
+					++primary_iter;
+					if (primary_iter != End)
+					{
+						if (HasComponents<SecondaryComponents...>(&*primary_iter->GetOwner()))
+							break;
+					}
+					else
+						break;
+				} while (true);
```

Comment:
Remove all ifs from do while and simply write:
`do { ++primary_iter; } while (primary_iter != End && !HasComponents<SecondaryComponents...>(&*primary_iter->GetOwner()));`

Do it similarly for decrement.